projects/TransFusion/transfusion/hungarian_assigner.py

```diff
@@ -0,0 +1,187 @@
+import torch
+from mmdet.models.task_modules import AssignResult, BaseAssigner
+from mmengine.structures import InstanceData
+
+from mmdet3d.registry import TASK_UTILS
+
+try:
+    from scipy.optimize import linear_sum_assignment
+except ImportError:
+    linear_sum_assignment = None
+
+
+@TASK_UTILS.register_module()
+class BBox3DL1Cost(object):
+
+    def __init__(self, weight):
+        self.weight = weight
+
+    def __call__(self, bboxes, gt_bboxes, train_cfg):
+        reg_cost = torch.cdist(bboxes, gt_bboxes, p=1)
+        return reg_cost * self.weight
+
+
+@TASK_UTILS.register_module()
+class BBoxBEVL1Cost(object):
+
+    def __init__(self, weight):
+        self.weight = weight
+
+    def __call__(self, bboxes, gt_bboxes, train_cfg):
+        pc_start = bboxes.new(train_cfg['point_cloud_range'][0:2])
+        pc_range = bboxes.new(
+            train_cfg['point_cloud_range'][3:5]) - bboxes.new(
+                train_cfg['point_cloud_range'][0:2])
+        # normalize the box center to [0, 1]
+        normalized_bboxes_xy = (bboxes[:, :2] - pc_start) / pc_range
+        normalized_gt_bboxes_xy = (gt_bboxes[:, :2] - pc_start) / pc_range
+        reg_cost = torch.cdist(
+            normalized_bboxes_xy, normalized_gt_bboxes_xy, p=1)
+        return reg_cost * self.weight
+
+
+@TASK_UTILS.register_module()
+class IoU3DCost(object):
+
+    def __init__(self, weight):
+        self.weight = weight
+
+    def __call__(self, iou):
+        iou_cost = -iou
+        return iou_cost * self.weight
+
+
+@TASK_UTILS.register_module()
+class HeuristicAssigner3D(BaseAssigner):
```

Comment:
Could reuse the implementation in BEVFusion and del this file?
```python
from ..BEVFusion.bevfusion.utils import HungarianAssigner3D
from ..BEVFusion.bevfusion.transformer import TransformerDecoderLayer  # noqa F401
from ..BEVFusion.bevfusion.utils import TransFusionBBoxCoder
```

Reply:
Sure. Is it better to merge [these changes](https://github.com/open-mmlab/mmdetection3d/pull/2558)?

Reply:
@JingweiZhang12 I pushed the code based on the original repository. I'm currently working on using BEVFusion implementation [here](https://github.com/wep21/mmdetection3d/tree/transfusion-with-bevfusion), but loss_bbox doesn't decrease compared to the original one.
I guess I made a mistake when integrating your code, so I appreciate it if you could give me some advice to resolve this issue.

Reply:
@wep21 Hi, the original BEVFusion has a potential bug: https://github.com/mit-han-lab/bevfusion/issues/419. This will cause the loss_bbox can not decrease. We have fixed it in the latest branch.

Reply:
@JingweiZhang12 I found out the changes from original code cause problem when training transfusion. I reverted the changes [here](https://github.com/open-mmlab/mmdetection3d/pull/2547/commits/2ad5e4ac9cbbfc2a93a99ca85983d57d7e05b450) and then loss_bbox and matched ious become sane.
Do you have any thoughts about this?

Reply:
Hi， this is because the voxelization output order of the transfusion is different from ours. TransFusion: zyx,   Ours: zxy

Reply:
It may not have an influence on the accuracy of the model. The style of the original Transfusion is the same as ours.

Reply:
@JingweiZhang12 I would like to use the same coordinate as original, so is it better to create another head implementation in transfusion?

Reply:
@wep21 Yep, it's better to use the original head implements of TransFusioin to align the training precision easily since the BEVFusion head may modify something and have some potential incompatibilities.